pallets/messages/src/lib.rs

```diff
@@ -0,0 +1,268 @@
+	use super::*;
+	use common_primitives::msa::MessageSenderId;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type WeightInfo: WeightInfo;
+
+		/// The maximum number of messages in a block
+		#[pallet::constant]
+		type MaxMessagesPerBlock: Get<u16>;
+
+		/// The maximum size of a message [Byte]
+		#[pallet::constant]
+		type MaxMessageSizeInBytes: Get<u32>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_block_messages)]
+	pub(super) type BlockMessages<T: Config> //TODO: convert Vec to BoundedVec
+		= StorageValue<_, Vec<(Message<T::AccountId>, SchemaId)>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_messages)]
+	pub(super) type Messages<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::BlockNumber,
+		Twox64Concat,
+		SchemaId,
+		Vec<Message<T::AccountId>>,
+		ValueQuery,
+	>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Too many messages are added to existing block
+		TooManyMessagesInBlock,
+		/// Message size is too large
+		TooLargeMessage,
+		/// Invalid Pagination Request
+		InvalidPaginationRequest,
+		/// Type Conversion Overflow
+		TypeConversionOverflow,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Messages are stored for a specified schema id and block number
+		MessagesStored { schema_id: SchemaId, block_number: T::BlockNumber, count: u16 },
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(current: T::BlockNumber) -> Weight {
+			let prev_block = current - T::BlockNumber::one();
+			Self::move_messages_into_final_storage(prev_block)
+			// TODO: add retention policy execution
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Adds a message into storage
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `schema_id`: Registered schema id for current message
+		/// - `message`: Serialized message data
+		///
+		/// Result is equivalent to the dispatched result.
+		///
+		/// # <weight>
+		/// Execution weight
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::add(message.len() as u32, 1_000))]
+		pub fn add(origin: OriginFor<T>, schema_id: SchemaId, message: Vec<u8>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				message.len() < T::MaxMessageSizeInBytes::get().try_into().unwrap(),
+				Error::<T>::TooLargeMessage
+			);
+			// TODO: validate schema existence and validity from schema pallet
+			// TODO: get MsaId from mrc-identity pallet
+
+			let mut block_messages = <BlockMessages<T>>::get();
+			let current_size: u16 = block_messages
```

Comment:
I think this needs to be moved this validation into Signed Extension. I think we can avoid having a user check how full a block is before actually executing.

Let me know if you wanna pair.